### ISSUE TITLE
LaTeX Beamer and Xournalpp

### ISSUE THREAD

**Issue by gyps** [xournalpp/xournalpp#1647]
It would be great if one could compile a latex beamer presentation into a xournalpp-file. With overlay slides as overlays in xournalpp. Since I don't know anything about the xournalpp file format I don't know if that would be hard to implement or even possible at all.

**Comment by LittleHuba** [comment]
We are currently working on a completely new fileformat. Follow our organisation on Github to get a notification once we release the repository for the new library.

What would most likely be the easiest way is to internally render the beamer presentation as PDF and then use it just as background. Is that what you would like to have or did I missunderstand you?

**Comment by dbosk** [comment]
Beamer overlays allows certain content on a slide to be hidden and shown as progressing through a slide. Like animations. Say that we have a bullet list. First we show the first bullet, whereas the second is hidden. Then we show the second, and so on. Beamer solves it by rendering the slide several times: one PDF page of the slide with the first visible, the second hidden; the following PDF page is the same slide but with the first bullet hidden and second shown.

My interpretation of his request is to detect this and keep them as the same slide. Thus any annotations made while the first bullet is shown, would still be visible when the second is shown instead. However, this would probably require Beamer to generate xournalpp-files instead of PDFs. Or adding some magic in xournalpp to detect this, similarly to how [pdfpc](https://pdfpc.github.io/) detects it.

**Comment by gyps** [comment]
> My interpretation of his request is to detect this and keep them as the same slide. Thus any annotations made while the first bullet is shown, would still be visible when the second is shown instead. 

That's exactly what I meant (thanks for clarifying)

> However, this would probably require Beamer to generate xournalpp-files instead of PDFs. Or adding some magic in xournalpp to detect this, similarly to how [pdfpc](https://pdfpc.github.io/) detects it.

Both options are possible but option 2 has probably the additional advantage that if xournalpp were aware of additional information provided within the PDF anotation could be linked to tags in the PDF so after recreating the PDF they would still be associated with the correct slides.

**Comment by LittleHuba** [comment]
Making Beamer generate xournalpp-files is not really a good option. The new format is not as simple as the current one. Also updating the slides would mean you loose all annotations. The new format also has the paradigm of not modifying attachments and stroing attachments in a way that they can be exchanged later on without issue. That is not really possible with this option.

Option two sounds better. We could add a parsing module for Beamer presentations, which could parse the raw latex internally and thus get the most information out of the format. We already have a latex dependency for our formula tool so this adds no further problems on that front. The parsing will be fairly complex though and I'd be a fan of outsourcing this module to a library that we can just include into the main project.

Also with our current team size this will take quite long. We are standing before a major poject overhaul which will take us quite some time. So if you know of anybody that would be interested in helping us, we would be very happy to take him on board. And chances that this feature is implemented earlier are increasing with every team member we gain.

**Comment by gyps** [comment]
Parsing the LaTeX code is probably unnecessary. All necessary information could be placed inside the PDF and parsed from there (libpoppler) already does this. To find out which pages are grouped together page numbers are already sufficient. And if one wants more as I suggested more information could be placed into the PDF. Similar to the hack I [did here](https://github.com/politza/pdf-tools/issues/243)  two years ago to show arbitrary information in emacs mode-line when presenting PDFs. This way one could tag PDF pages and link the xournalpp-annotations to exectly that page so if the PDF is updated, annotations will still refer to the correct page.

**Comment by LittleHuba** [comment]
Would be a valid option as well, but slightly more work for users if I see it correctly. You have to manually add the information to the PDF I assume?

**Comment by timtro** [comment]
> Would be a valid option as well, but slightly more work for users if I see it correctly. You have to manually add the information to the PDF I assume?

I don't think so. The pdf pages belonging to the same "frame" of the presentation have the same "page name" (or whatever the PDF jargon is, "Page 1", etc). I think all this requires is a condition that annotations are kept as long as the pdf "page name" doesn't change.

**Comment by torwag** [comment]
@timtro I think that the PDF page name changes. What beamer does is actually to create a single PDF page for each animation step. What is needed is to read in e.g. the page number and as long as this does not change it is actually the same beamer frame.
Another route could be the usage of a counter within beamer and to store this counter with each and every pdf page in an external file. That can be done in LaTeX (albeit, it requires to modify the beamer class or to introduce a small new helper class). In fact LaTeX creates quit a few very similar files during the compilation already. The PDF together with this external file could be read in by xournal. The file could be as simple as 
```
1,1-4
2,5-6
3,7
4,8-34
```
read: frame1 consists of pdf pages 1-4, frame2 of pdf pages 5 and 6.... etc.

This solution has several advantages:
1. It is rather light weight on the xournalpp side. Check if an additional file is available and if parse it. Modify xournalpp to keep annotations on pages which belongs to the same group.
2. The generation of this file is not restricted to latex beamer. Other tools could be used to generate this file as well. Even hand-made would be an option and actually does not take so much time. That would enable to convert old beamer-based presentations (or other PDF presentations). 
3. A small script could be used to generate alternatively such a file automatically based on different criteria (like the page-number), if there is no way to generate it during the PDF generation. That would be external and outside xournalpp and outside LaTeX-beamer. Which removes workload from both projects.

**Comment by gyps** [comment]
That was more or less my idea but instead of using an external file I thought about using invisible PDF comments (to avoid the second file). I also thought about creating a new hash for every frame/page the first time it is created. So the user need not deal with giving the pages names and at the same time the hash could persistently identify each page so comments could refer to it.

**Comment by torwag** [comment]
Well embedding the information about the slides inside the PDF is for sure the better solution from the user perspective. However, it requires a modification of beamers workflow or as @gyps  mentioned the user to create certain manually notes or something similar. Beamer+LaTeX is powerful enough to allow to create a file or a note automatically. Thus, it might be possible to provide a code snippet which might later be integrated into beamer.
Pro for an external file would be that it can be used for many other PDF sources as well.

**Comment by rolandlo** [comment]
In LaTeX Beamer the same content will not necessarily be displayed at the same position of the slide.

For instance if you have something like
\only<1>{content for slide 1 only}
\only<1-2> {shared content}
then the shared content will be displayed on different positions on the two pdf-pages. On the first page it is after the contents for slide 1 only, on the second page it is the first thing displayed. 

It would be hard to match the positions, where content is displayed on the different pages, and tracking which annotations are linked to which (shared) content. Hence I think full support of LaTeX Beamer overlays will be pretty difficult.

**Comment by flo2k** [comment]
@rolandlo for sure, this would not be easy.

In my case I do only use the various uncover features of the latex beamer package to add things and to not move it around. I think a lot of other poeple also only add things without changing the position of elements. So, a first solution that does this positioning problem do not take into account would--for me--be a possible solution.

**Comment by LittleHuba** [comment]
The only option I see is to actually use the Latex file itself as attachment and use that to determine all that stuff. We can then take latex to compile the attachment internally to a pdf and display it while still knowing what the content will be. That would still mean we need to parse the latex file though... 

All in all very complex and something our team would definitely need help with to get it working.

**Comment by flo2k** [comment]
I think its not a good idea to use the latex file directly. In my case, I have "compiler switches" (I define commands from the external) to decide which version of the slides to build: one version for the presentation with all that animations, the other without e.g. for printing.

And then consider the different commands pdflatex, xetex, ... that can be used.

I think working on the PDF is the better way.

**Comment by atticus-sullivan** [comment]
Just stumbled over this. One way would be to just increment/decrement the pdf page a xopp page is showing (at least as long as we don't care about tracking shared content like you described).

Not sure if we can access pdf metadata (like page numbers) from the plugin system. If we can, it should be not that hard to implement three plugin functions:
- condense the document to only contain one xopp page for every slide
- step forward, check if on last animation step of the slide -> go to next xopp page / increment pdf background page
- step backward, analog to the above one

**Comment by timtro** [comment]
> I think its not a good idea to use the latex file directly. In my case, I have "compiler switches" (I define commands from the external) to decide which version of the slides to build: one version for the presentation with all that animations, the other without e.g. for printing.
> 
> And then consider the different commands pdflatex, xetex, ... that can be used.
> 
> I think working on the PDF is the better way.

I wish I had more time to look into this. From what I recall, other tools I have seen (like ['impressive'](https://impressive.sourceforge.net/) pdf presenter tool) can take beamer's output from the `.aux` file generated during LaTeX compilation. In the case of 'impressive', I believe this is done so that the `<tab>` overview doesn't generate a thumbnail for every transition within a `frame`. I think it's an external tool that generates a configuration file for impressive, not impressive itself.

If someone is working on this, I'm happy to look into it a bit more and collaborate on a PR.

**Comment by atticus-sullivan** [comment]
@timtro feel free to look into the PR linked right above (#5837). The approach based on the page numbers/labels doesn't sound too different compared to what you describe of impressive's functionality.
